// @swimlane/ngx-dnd/esm5/lib/ngx-dnd.module.js
import { __decorate, __read, __spread } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DraggableDirective } from './directives/ngx-draggable.directive';
import { DroppableDirective } from './directives/ngx-droppable.directive';
import { DragHandleDirective } from './directives/ngx-drag-handle.directive';
import { ContainerComponent } from './components/container/container.component';
import { ItemComponent } from './components/item/item.component';
import { DrakeStoreService } from './services/drake-store.service';
var components = [ContainerComponent, ItemComponent];
var directives = [DraggableDirective, DroppableDirective, DragHandleDirective];
var NgxDnDModule = /** @class */ (function () {
    function NgxDnDModule() {
    }
    NgxDnDModule_1 = NgxDnDModule;
    NgxDnDModule.forRoot = function () {
        return {
            ngModule: NgxDnDModule_1,
            providers: [DrakeStoreService]
        };
    };
    var NgxDnDModule_1;
    NgxDnDModule = NgxDnDModule_1 = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: __spread(components, directives),
            exports: __spread(components, directives)
        })
    ], NgxDnDModule);
    return NgxDnDModule;
}());
export { NgxDnDModule };
