// @swimlane/ngx-dnd/esm2015/lib/ngx-dnd.module.js
var NgxDnDModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DraggableDirective } from './directives/ngx-draggable.directive';
import { DroppableDirective } from './directives/ngx-droppable.directive';
import { DragHandleDirective } from './directives/ngx-drag-handle.directive';
import { ContainerComponent } from './components/container/container.component';
import { ItemComponent } from './components/item/item.component';
import { DrakeStoreService } from './services/drake-store.service';
const components = [ContainerComponent, ItemComponent];
const directives = [DraggableDirective, DroppableDirective, DragHandleDirective];
let NgxDnDModule = NgxDnDModule_1 = class NgxDnDModule {
    static forRoot() {
        return {
            ngModule: NgxDnDModule_1,
            providers: [DrakeStoreService]
        };
    }
};
NgxDnDModule = NgxDnDModule_1 = __decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [...components, ...directives],
        exports: [...components, ...directives]
    })
], NgxDnDModule);
export { NgxDnDModule };
