// @swimlane/ngx-dnd/esm2015/lib/directives/ngx-droppable.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, Output, OnInit, OnDestroy, AfterViewInit, ElementRef, EventEmitter, Renderer2 } from '@angular/core';
import { DrakeStoreService } from '../services/drake-store.service';
let i = 10000;
function getNextId() {
    return i++;
}
/**
 * Makes the container droppable and children draggable.
 *
 * @export
 */
let DroppableDirective = class DroppableDirective {
    constructor(el, renderer, drakesService) {
        this.el = el;
        this.renderer = renderer;
        this.drakesService = drakesService;
        this.copy = false;
        this.removeOnSpill = false;
        this.direction = 'vertical';
        this.drop = new EventEmitter();
        this.drag = new EventEmitter();
        this.over = new EventEmitter();
        this.out = new EventEmitter();
        this.remove = new EventEmitter();
        this.cancel = new EventEmitter();
    }
    get container() {
        return this.el.nativeElement;
    }
    get dropZone() {
        return this._dropZone || this.ngxDroppable || this.defaultZone;
    }
    set dropZone(val) {
        this._dropZone = val;
    }
    ngOnInit() {
        this.defaultZone = `@@DefaultDropZone-${getNextId()}@@`;
        this.drakesService.register(this);
    }
    ngAfterViewInit() {
        this.over.subscribe(() => {
            this.renderer.addClass(this.container, 'gu-over');
        });
        this.out.subscribe(() => {
            this.renderer.removeClass(this.container, 'gu-over');
        });
    }
    ngOnDestroy() {
        this.drakesService.remove(this);
    }
};
DroppableDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: DrakeStoreService }
];
__decorate([
    Input(),
    __metadata("design:type", Object)
], DroppableDirective.prototype, "model", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], DroppableDirective.prototype, "copy", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], DroppableDirective.prototype, "removeOnSpill", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], DroppableDirective.prototype, "ngxDroppable", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], DroppableDirective.prototype, "direction", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], DroppableDirective.prototype, "drop", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], DroppableDirective.prototype, "drag", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], DroppableDirective.prototype, "over", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], DroppableDirective.prototype, "out", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], DroppableDirective.prototype, "remove", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], DroppableDirective.prototype, "cancel", void 0);
__decorate([
    Input(),
    __metadata("design:type", String),
    __metadata("design:paramtypes", [String])
], DroppableDirective.prototype, "dropZone", null);
DroppableDirective = __decorate([
    Directive({ selector: '[ngxDroppable]' }),
    __metadata("design:paramtypes", [ElementRef, Renderer2, DrakeStoreService])
], DroppableDirective);
export { DroppableDirective };
