// @swimlane/ngx-dnd/esm2015/lib/components/container/container.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, ViewEncapsulation, ContentChild, TemplateRef, ViewChild, EventEmitter } from '@angular/core';
import { DroppableDirective } from '../../directives/ngx-droppable.directive';
let i = 0;
function getNextId() {
    return i++;
}
/**
 * Component that allows nested ngxDroppable and ngxDraggables
 *
 * @export
 */
let ContainerComponent = class ContainerComponent {
    constructor() {
        this.copy = false;
        this.removeOnSpill = false;
        this.dropZone = `@@DefaultDropZone-${getNextId()}@@`;
        this.drop = new EventEmitter();
        this.drag = new EventEmitter();
        this.over = new EventEmitter();
        this.out = new EventEmitter();
        this.remove = new EventEmitter();
        this.cancel = new EventEmitter();
    }
    get dropZones() {
        return this._dropZones || this._defaultZones;
    }
    set dropZones(val) {
        this._dropZones = val;
    }
    // @Input() classes: any = {};
    // @Input() dragulaOptions: any;
    set templateInput(template) {
        this.template = template;
    }
    set templateChild(template) {
        this.template = template;
    }
    ngOnInit() {
        this._defaultZones = [this.dropZone];
    }
    ngAfterViewInit() {
        this.droppable.drag.subscribe((v) => this.drag.emit(v));
        this.droppable.drop.subscribe((v) => this.drop.emit(v));
        this.droppable.over.subscribe((v) => this.over.emit(v));
        this.droppable.out.subscribe((v) => this.out.emit(v));
        this.droppable.remove.subscribe((v) => this.remove.emit(v));
        this.droppable.cancel.subscribe((v) => this.cancel.emit(v));
    }
};
__decorate([
    Input(),
    __metadata("design:type", Object)
], ContainerComponent.prototype, "model", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], ContainerComponent.prototype, "copy", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], ContainerComponent.prototype, "removeOnSpill", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], ContainerComponent.prototype, "droppableItemClass", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], ContainerComponent.prototype, "dropZone", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], ContainerComponent.prototype, "dropZones", null);
__decorate([
    Input(),
    __metadata("design:type", Function)
], ContainerComponent.prototype, "moves", void 0);
__decorate([
    Input('template'),
    __metadata("design:type", TemplateRef),
    __metadata("design:paramtypes", [TemplateRef])
], ContainerComponent.prototype, "templateInput", null);
__decorate([
    ContentChild(TemplateRef, { static: true }),
    __metadata("design:type", TemplateRef),
    __metadata("design:paramtypes", [TemplateRef])
], ContainerComponent.prototype, "templateChild", null);
__decorate([
    ViewChild(DroppableDirective, { static: true }),
    __metadata("design:type", Object)
], ContainerComponent.prototype, "droppable", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], ContainerComponent.prototype, "drop", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], ContainerComponent.prototype, "drag", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], ContainerComponent.prototype, "over", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], ContainerComponent.prototype, "out", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], ContainerComponent.prototype, "remove", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], ContainerComponent.prototype, "cancel", void 0);
ContainerComponent = __decorate([
    Component({
        selector: 'ngx-dnd-container',
        template: "<div\n  ngxDroppable\n  [dropZone]=\"dropZone\"\n  [model]=\"model\"\n  [copy]=\"copy\"\n  [ngClass]=\"{ 'gu-empty': !model?.length }\"\n  [removeOnSpill]=\"removeOnSpill\"\n  class='ngx-dnd-container'>\n  <ng-container *ngIf=\"model\">\n    <ng-container *ngFor=\"let item of model\">\n      <ngx-dnd-item\n        ngxDraggable\n        [model]=\"item\"\n        [dropZone]=\"dropZone\"\n        [dropZones]=\"dropZones\"\n        [copy]=\"copy\"\n        [moves]=\"moves\"\n        [removeOnSpill]=\"removeOnSpill\"\n        [droppableItemClass]=\"droppableItemClass\">\n      </ngx-dnd-item>\n    </ng-container>\n  </ng-container>\n  <ng-content *ngIf=\"!model\"></ng-content>\n</div>\n",
        encapsulation: ViewEncapsulation.None,
        styles: [".ngx-dnd-container{background-color:rgba(255,255,255,.2);border:2px solid red;margin:10px;padding:10px}.ngx-dnd-container.gu-empty{border:2px dotted red}.ngx-dnd-container:nth-child(odd){background-color:rgba(0,0,0,.2)}.ngx-dnd-container .ex-moved{background-color:#e74c3c}.ngx-dnd-container .ex-over{background-color:rgba(255,255,255,.3)}.ngx-dnd-container .handle{padding:0 5px;margin-right:5px;background-color:rgba(0,0,0,.4);cursor:move}.no-select{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.clearfix::after{content:\" \";display:block;height:0;clear:both}"]
    })
], ContainerComponent);
export { ContainerComponent };
